src/authorized_keys_d/as_user/as_user.c

```diff
@@ -0,0 +1,205 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <sched.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "as_user.h"
+
+// By doing these operations in CGO, we prevent the Go scheduler from
+// potentially executing while our uid/gid/umask is changed.
+//
+// By doing them in a new !CLONE_FS thread we also localize the effects of the
+// umask() switching.
+//
+// The calling go code must supply the appropriate args struct.
+// We create a new thread passing the args down verbatim, it's all fairly
+// mechanical.  There's no need to restore umask/eids since we just exit the
+// created thread when done.
+//
+// The return value and errno is simply fed back to the caller from the thread
+// through some variables synchronized using a closed pipe.
+
+#define STACK_SIZE (64 * 1024)
+
+typedef struct au_thread_ctxt {
+	au_ids_t	*ids;
+
+	char		*stack;
+	int		(*fn)(void *);
+	void		*fn_args;
+
+	int		ret;
+	int		err;
+
+	int		fin_pipe[2];
+} au_thread_ctxt_t;
+
+/* set_eids() sets the effective gid and uid of the calling thread to ids */
+static int set_eids(au_ids_t *ids) {
+	uid_t	cu;
+	gid_t	cg;
+
+	umask(077);
+
+	cu = geteuid();
+	cg = getegid();
+
+	if(cg != ids->gid && setregid(-1, ids->gid) == -1)
```

Comment:
is there any reason to do that?  this isn't portable code calling clone etc.  won't glibc just turn those into setre[ug]id() calls anyways?